### PR TITLE
gbXML External Reference

### PR DESCRIPTION
**Pull request by nllong**
Add support for referencing gbXML elements from within BuildingSync

### CODE REVIEW

**Comment by nllong** [comment]
@aqwolfe -- can you resolve the conflicts please?

**Comment by nllong** [comment]
Resolve differences between 
* https://github.com/BuildingSync/schema/pull/156/files#diff-04c6e90faac2675aa89e2176d2eec7d8
* https://github.com/BuildingSync/schema/pull/60/files#diff-07df69262a9c145a5a91343bc454ab34

Favor the changes that are in this PR; however, pull in the references from the other PR.

File: proposals/2018/Rename Root Element.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ Standard 211-type audit. For example:
 4) transferring building data for use in M&V IPMVP Option C/D type analysis
```
Comment (nllong) [inline proposals/2018/Rename Root Element.md]: We don't want to update the proposal from 2018/Rename Root Element. This should be a new proposal in 2019 for External References.

**Comment by dnicholaspsd** [comment]
@nllong I've resolved the conflict with the 'Rename Root Element' proposal and added the external reference proposal from PR 60.  I've also manually resolved the conflicts with the BuildingSync.xsd file so it should be able to merge easily.  It's still listed on the 'Conflicting files' tab below and when I look at Github's 'Resolve conflicts' UI, it seems to be showing the file differences from the earlier commit of this branch.  I don't know enough about Github's pull request system to know exactly why, but perhaps it has something to do with the Travis CI Build failing.  When I look at that and research the error message from it, it seems to have to do with the BuildingSync.xsd file referencing a schema that isn't present (GBXML).  Not sure how you'd like to proceed.  Include the GBXML schema in the repo?  Somehow change the build check process so it doesn't fail without the referenced schema?

**Comment by dnicholaspsd** [comment]
Merge conflicts have been resolved, just need to deal with the build failures.

**Comment by nllong** [comment]
Thanks @dnicholaspsd for fixing the merge issue. Looks like Nokogiri isn't happy with the XSD format now.

```
 Nokogiri::XML::SyntaxError:
       1851:0: ERROR: Element '{http://www.w3.org/2001/XMLSchema}element', attribute 'ref': The QName value '{http://www.gbxml.org/schema}SpaceBoundary' does not resolve to a(n) element declaration.
```

**Comment by nllong** [comment]
@dnicholaspsd @aqwolfe -- have you had a chance to look at resolving this issue:

```
Failure/Error: @xsd = Nokogiri::XML::Schema(File.read('BuildingSync.xsd'))
     
     Nokogiri::XML::SyntaxError:
       1851:0: ERROR: Element '{http://www.w3.org/2001/XMLSchema}element', attribute 'ref': The QName value '{http://www.gbxml.org/schema}SpaceBoundary' does not resolve to a(n) element declaration.
```

**Comment by dnicholaspsd** [comment]
@nllong I have; I sent you an email about it, but it must have gotten lost in the wind.  The build is failing because the updated BuildingSync schema is trying to reference the GBXML schema, but that file isn’t present in the repo, because it was decided it was preferable not to have it there and have to keep it updated, etc.  The only thing I can do on my end to fix this build error is to add that file to the repo.  If that’s acceptable for you, I can do that.  If not, the only ideas I can come up with for you (I don’t know the feasibility of them) are to edit the build process somehow to not fail on this error, or to somehow make the schema file available to the build process (so it succeeds), but not part of the actual repo.

**Comment by nllong** [comment]
thanks for the update. So it sounds like we will need to have the gbxml schema to be included with BSync anytime we want to validate, right? Is there no way to reference it by URI in the header of the BSync schema?

**Comment by dnicholaspsd** [comment]
@nllong Apparently there is a way to do just that.  Looks like everything went well with the validation this time around.

**Comment by nllong** [comment]
@markborkum, at the moment this is the full extent, but I guarantee that there will be other elements that we will want to pull in. Part of the use case is to support ASHRAE bEQ upload with some gbXML mixed into BuildingSync. 

I propose that we merge this down and then reach out to gbXML. Let me know your thoughts.

**Comment by markborkum** [comment]
@nllong I'm hesitant to merge until we do an assessment of possible name collisions (e.g., a script that constructs the set of element/type names in each schema document and then tests for intersection).

**Comment by nllong** [comment]
We are only supporting a subset of gbxml (namely Space). I agree that we should do a name collision check. We can work on that before merging.


Il giorno 17 nov 2019, alle ore 12:31, Mark Borkum <notifications@github.com> ha scritto:

﻿

@nllong<https://github.com/nllong> I'm hesitant to merge until we do an assessment of possible name collisions (e.g., a script that constructs the set of element/type names in each schema document and then tests for intersection).

**Comment by markborkum** [comment]
@nllong Capacity, Latitude, Location, Longitude, Manufacturer, State, and ZipCode.

**Comment by markborkum** [comment]
> @nllong Capacity, Latitude, Location, Longitude, Manufacturer, State, and ZipCode.

The top-level elements that are shared by both schemas, which would result in name collisions when using the `xsd2ruby.rb` script.

```ruby
#!/usr/bin/env ruby -Wall
# frozen_string_literal: true

require "set"

require "nokogiri"

FILENAMES = %w(BuildingSync.xsd GreenBuildingXML_Ver6.01.xsd)

NAMESPACES = {
  xs: "http://www.w3.org/2001/XMLSchema",
}

NAME_ATTRIBUTE = "name"

XPATH_FOR_NAME_ATTRIBUTE = %w(attribute complexType element simpleType).collect { |s|
  "./xs:schema/xs:#{s}[@#{NAME_ATTRIBUTE}]"
}.join(" | ")

sets = FILENAMES.collect { |filename|
  nokogiri_xml_document = File.open(filename, "r") do |file|
    Nokogiri::XML(file)
  end

  nokogiri_xml_document.xpath(XPATH_FOR_NAME_ATTRIBUTE, **NAMESPACES).collect { |nokogiri_xml_node|
    nokogiri_xml_node.attribute(NAME_ATTRIBUTE).value
  }.to_set
}

sets.each_with_index do |orig_set, orig_index|
  unless orig_index == 0
    sets.each_with_index do |new_set, new_index|
      unless orig_index == new_index
        intersection_set = orig_set.intersection(new_set)

        $stdout.puts("intersection(#{FILENAMES[orig_index].inspect}, #{FILENAMES[new_index].inspect}) = #{intersection_set.to_a.sort.inspect}")
      end
    end
  end
end
```

```bash
$ ./app.rb 
intersection("GreenBuildingXML_Ver6.01.xsd", "BuildingSync.xsd") = ["Capacity", "Latitude", "Location", "Longitude", "Manufacturer", "State", "ZipCode"]
```

**Comment by nllong** [comment]
Besides the fact that this PR has conflicts--do you think we can merge this down for this month's release?  

The hesitation was documenting the potential conflicts which includes `["Capacity", "Latitude", "Location", "Longitude", "Manufacturer", "State", "ZipCode"]`.  And I think we need to add the script above as a rake task to catch any potential changes.

Thoughts @markborkum @corymosiman12 @macintoshpie

**Comment by markborkum** [comment]
> The hesitation was documenting the potential conflicts which includes `["Capacity", "Latitude", "Location", "Longitude", "Manufacturer", "State", "ZipCode"]`. And I think we need to add the script above as a rake task to catch any potential changes.

Basically, given the implementation of the `xsd2ruby.rb` script that Audit Template, the hesitation is any situation where we have:

```xml
<auc:Parent>
  <auc:Child/>
  <gbxml:Child/>
</auc:Parent>
```

The `xsd2ruby.rb` script, which uses a templating system that not account for XML namespaces prefixes, would generate something like:

```ruby
# auc:Parent
class Parent
  attr_accessor :child, :child

  # child - auc:Child
  # child - gbxml:Child
  def initialize(child = nil, child = nil)
    @child = child
    @child = child
  end
end

# auc:Child
class Child
end

# gbxml:Child
class Child
end
```

This point would be a blocker for Audit Template (and other code generation tools that do not account for XML namespace prefixes).

**Comment by macintoshpie** [comment]
@nllong @markborkum I've been debugging the failing test, it looks like Nokogiri doesn't like having a url for the imported gbxml `schemaLocation`. I was able to get the tests passing by downloading the xsd for gbxml, then providing a full path to the schema, but obviously keeping their schema in our repo is not ideal. Is there something I'm missing about Nokogiri which would allow it to handle this?

If not, I'm thinking our only other option is to download the schema temporarily and replace the schemaLocation with the local path to the file, but that also is less than ideal. Do yall have any other ideas?

**Comment by nllong** [comment]
For running the tests then downloading temporarily and changing the schemaLocation would be okay. This would happen all within the rspec and shouldn't impact the user's interaction. The drawback though is if anyone else is using Nokogiri with BuildingSycn schema then it won't work.

**Comment by macintoshpie** [comment]
> The drawback though is if anyone else is using Nokogiri with BuildingSycn schema then it won't work.
Yeah, that's a bummer but more of a Nokogiri issue than BuildingSync, so I guess this isn't too bad

File: spec/validation_spec.rb [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
   before :all do
-    @xsd = Nokogiri::XML::Schema(File.read('BuildingSync.xsd'))
+    # Nokogiri doesn't seem to support XSDs which import other schemas with URLs
+    # for the schemaLocation. To allow testing, we download the imported schema
+    # and point schemaLocation to it instead
+    schema_doc = Nokogiri::XML(File.read('BuildingSync.xsd'))
+
+    GEOJSON_XSD_PATH = 'geojson.xsd'
+    GEOJSON_IMPORT_XPATH = 'xs:schema/xs:import[@namespace = "http://www.gbxml.org/schema"]'
+    if !File.file?(GEOJSON_XSD_PATH) then
```
Comment (nllong) [inline spec/validation_spec.rb:15]: thanks for adding this hack!

File: Rakefile [hunk 3]
```diff
@@ -84,6 +86,63 @@ task :update_documentation do
   File.open("BuildingSync.xsd", 'w') { |f| f << doc.to_xml(:indent => 2) }
 end
 
+desc 'Get report of naming collisions between schemas'
+task :check_collisions do
```
Comment (nllong) [inline Rakefile:90]: @macintoshpie -- can you have this run within an rspec. Probably create a new file called check_collisions_spec.rb. I may be missing something here though.

File: Rakefile [hunk 4]
```diff
@@ -84,6 +86,63 @@ task :update_documentation do
...
+      end
+    end
+
+    nokogiri_xml_document.xpath(XPATH_FOR_NAME_ATTRIBUTE, **NAMESPACES).collect { |nokogiri_xml_node|
+      nokogiri_xml_node.attribute(NAME_ATTRIBUTE).value
+    }.to_set
+  }
+
+  # permitted because AT’s xsd2ruby script has already been tested with them and it was OK
+  PERMITTED_CONFLICTS = %w(Capacity Latitude Location Longitude Manufacturer State ZipCode)
```
Comment (nllong) [inline Rakefile:126]: 👍

**Review by markborkum — APPROVED** [review]
Re-tested `BuildingSync.xsd` using the Audit Template `xsd2ruby.rb` code generation pipeline to check for gbXML-related regressions.

Changes look good, thanks.

File: spec/validation_spec.rb (hunk 2 above)
Comment (markborkum) [reply spec/validation_spec.rb:15]: For backend code, where an XSD file is used, but the end-user never sees the XSD file directly (such as the `BuildingSync.xsd` file in Audit Template's backend, which is used by the APIs for validation), I like to cache local copies of the XSD file and its `xs:import`s (and the imports of the imports, etc.), and then to change all of the `schemaLocation` attribute values from URLs to paths.

For example, in Audit Template's `xsd2ruby.rb` code generation pipeline, the gbXML import would be changed from

```
<xs:import namespace="http://www.gbxml.org/schema" schemaLocation="https://www.gbxml.org/schema/6-01/GreenBuildingXML_Ver6.01.xsd"/>
```

to

```
<xs:import namespace="http://www.gbxml.org/schema" schemaLocation="./GreenBuildingXML_Ver6.01.xsd"/>
```

Comment (markborkum) [reply spec/validation_spec.rb:15]: Also, note the weird "file:///F:/Temp/GreenBuildingXML_Ver6.01.xsd" value for the `schemaLocation` attribute in the `GreenBuildingXML_Ver6.01.xsd` file.

**Comment by nllong** [comment]
@markborkum - line showing that collision check ran on Travis: https://travis-ci.org/github/BuildingSync/schema/builds/728125476#L393

Thanks everyone for all the help getting this pushed through!

**Pull request merged**